docs/src/references/operation-workflow.md

```diff
@@ -0,0 +1,372 @@
+    let an agent operator override the automatic transition, say to check the downloaded file *before* moving the `Install` state.
+  - The `Init` state should be a *no-op* state.
+
+### Operation Implementation
+
+A workflow implementation for a specific operation must implement the actions specified for each non *no-op* state.
+This implementation has some freedom, notably to implement extra checks and actions but also to add new states.
+
+However, there are some rules and best practices.
+
+- All the state messages must be published as retained with QoS 1.
+- A workflow implementation should not reject a state message payload with unknown fields.
+  - It's also important to keep these unknown fields in the following states.
+  - This is important as we want to *extend* the workflow of an operation.
+    A software component added by the user might need these *extra* fields the plugin is not aware of.
+- A workflow implementation must not react on *no-op* states nor terminal states.
+  - The transition from a *no-op* state must be handled either by thin-edge as direct transition
+    or overridden by the user with domain-specific checks and actions.
+  - The terminal states, a.k.a `Successful` and `Failed`, are owned by the process which created the `Init` state (in practice, the mapper).
+    Only this process should clear the retained message state for an operation instance by sending an empty payload on command's topic.
+
+### Workflow Overriding
+
+Thin-edge provides a mechanism to override, extend and combine workflows.
+
+This mechanism is provided by the `tedge-agent` which gather a set of user-defined workflows
+and combined them with the builtin workflows implemented by the agent itself and the thin-edge operation plugins.
+
+Each workflow is defined using a TOML file that specifies:
+- the operation target and name defined as a topic filter
+  such as `te/device/main///cmd/software-update/+` or `te/device/main/service/+/cmd/restart/+`
+- the list of states
+- for each state:
+  - the state name as defined by the operation API
+  - the name of the state owner which is the process that is responsible for advancing a command at this state
+    (the default being `tedge`, meaning that thin-edge is responsible for running the state checks and actions)
+  - the set of states which can be an outcome for this state actions
+  - possible extra instructions on how to process the command at this stage, e.g.
+    - run a script
+    - emit an event
+    - raise an alarm
+  
+```toml title="file: workflow-example.toml"
+target = "te/device/main///cmd/configuration-update/+"
+
+[Init]
+  next = ["Download", "Failed"]
+  script = "/bin/schedule-configuration-update.sh"
+    
+[Download]
+  owner = "tedge-configuration-plugin"
+  next = ["Downloaded", "Failed"]
+
+[Downloaded]
+  script = "/bin/check-configuration-file.sh"
+  next = ["Install", "Failed"]
+
+[Install]
+  owner = "tedge-configuration-plugin"
```

Comment:
Provided `tedge-configuration-plguin` is a daemon, this `owner` field is just indicating the daemon name for logging purpose or like that? I assume that the state machine doesn't really care about the process owner?

Reply:
Something still needs to be clarified.

Indeed, the `tedge-configuration-plugin` can be implemented as a:
- as a daemon - as of today
- as a script - as you suggested [here](https://github.com/thin-edge/thin-edge.io/pull/2071#discussion_r1268196345)
- as an actor running inside the agent - as implemented by the [POC](https://github.com/didier-wenzek/tedge-mqtt-state-machine)

These 3 alternatives have different pros/cons as well as implementation impacts. I will dive into these.